### DISCUSSION TITLE
What the use case of the `ignore` property of the `onClickOutside`?

### DISCUSSION THREAD

**Discussion by Ky6uk** [vueuse/vueuse#2400] [General]
Hey. Recently I've tried to use `onClickOutside` as its component variant and have found that the component has `ignore` property that, according to the [documentation](https://vueuse.org/core/onclickoutside/#type-declarations) should work as **"List of elements that should not trigger the event."**.

That the exactly what I was looking for and have tried to pass an element to the `ignore` array, but have found that the event is still triggering. After some trying and findings I've found that some time ago the ignoring logic [has been removed](https://github.com/vueuse/vueuse/pull/2198/files#diff-8b308f45dba561e732d94afb100d76cc2aee8b7d265690c72e6fdf1e9dc59925L57-L63) from the main listener. The first obvious action that I've found is to create [the bug report](https://github.com/vueuse/vueuse/issues/2398) to make `ignore` work as intended and documented but [got a response](https://github.com/vueuse/vueuse/issues/2398#issuecomment-1302455106) that it shouldn't work.

To handle this case I've just returned the logic that has been removed to my local component inside the `@trigger` callback and everything became worked as expected again. But I'm a bit lost with the original possibility. Maybe I just don't understand how the `ignore` should work and maybe it's a documentation bug, lack of clarification or so. Thanks.

**Comment by curtgrimes** [comment]
Hi there, I took a look at [the `onClickOutside` docs](https://vueuse.org/core/onclickoutside/) and #2398 and I hope I can clear up some confusion.

It does look like the `ignore` option works correctly in the composable version of `onClickOutside()`. [Here is a POC of it](https://stackblitz.com/edit/vue3-script-setup-with-vite-zxxvth?file=src%2FApp.vue).

However, currently the [component version `<OnClickOutside/>`](https://github.com/vueuse/vueuse/blob/c5e91fbba264e9850cc4c3df2bfb96d907eedcf5/packages/core/onClickOutside/component.ts) does not support the same `ignore` option.

[This comment](https://github.com/vueuse/vueuse/issues/2398#issuecomment-1302455106) was correct but I think the further discussion about handling the click event in your specific POC might have confused the issue more. The docs at the moment don't mention this difference between `onClickOutside()` and `<OnClickOutside/>`, which I think explains your confusion.

I think we have two options here:
1. Update the docs to say the component version doesn't support `ignore` (and whatever else the component version doesn't support that `onClickOutside()` does), or
2. Make `<OnClickOutside/>` support the same things that `onClickOutside()` does. I don't know exactly what this would look like in component form or what precedent VueUse already has for components aligning with their composable forms. I think this option is preferable.  

I think another thing worth addressing is adding a demo of the `ignore` option, perhaps similar to the POC above.

@sibbng closed #2398 as "not planned", though, so I don't know if this difference between `onClickOutside()` and `<OnClickOutside/>` is intended or if PRs or further discussion on possibly implementing option 2 above are welcome.

**Reply by sibbng** [reply]
Thanks for clarifying the situation. Of course `OnClickOutside` component should have the options `onClickOutside` has. I just merged https://github.com/vueuse/vueuse/pull/2391 to support it.

I close the issue because the use case of the `ignore` option was misunderstood. Maybe we should document the configuration options like https://vueuse.org/core/useStyleTag/. PRs welcome 🙏

**Reply by Ky6uk** [reply]
It's working [as expected](https://stackblitz.com/edit/vitejs-vite-zte6t3/?file=src%2FApp.vue) for the composable function, so as a solution the component `OnClickOutside` could just start accepting the `ignore` property too and passing it further to the composable `onClickOutside(target, handler, { ignore: props.ignore })`.

It should work exactly as described in the documentation, the handler will not be fired in case the event has been triggered for one of the elements in the `ignore` list. The potential problem here might be reactivity in case the `ignore` array was changed, so additional work might be required also here.

**Reply by Ky6uk** [reply]
I just checked #2391 and I think it will fix the issue, so both of the variants should work as expected and documented.

1. [`onClickOutside()`](https://stackblitz.com/edit/vitejs-vite-zte6t3/?file=src%2FApp.vue) — currently working as expected and documented
2. [`<OnClickOutside>`](https://stackblitz.com/edit/vitejs-vite-djsqxz/?file=src%2FApp.vue) — doesn't work for `v9.4.0`, but should work in a new release.

Not sure it will handle reactivity in case `ignore` is reactive, but it sounds like another issue. 🤔

**Comment by Ky6uk** [comment]
FYI: yes, [`<OnClickOutside>`](https://stackblitz.com/edit/vitejs-vite-aymtpf/?file=src%2FApp.vue) now works as documented in `v9.5.0`.
